dbt/include/clickhouse/macros/materializations/incremental.sql

```diff
@@ -234,3 +244,36 @@
     {% do adapter.drop_relation(new_data_relation) %}
     {{ drop_relation_if_exists(distributed_new_data_relation) }}
 {% endmacro %}
+
+{% macro clickhouse__incremental_insert_replace(existing_relation, intermediate_relation, partition_by) %}
+    {% set new_data_relation = existing_relation.incorporate(path={"identifier": model['name']
+       + '__dbt_new_data_' + invocation_id.replace('-', '_')}) %}
+    {{ drop_relation_if_exists(new_data_relation) }}
+    {% call statement('create_new_data_temp') -%}
+      {{ get_create_table_as_sql(False, new_data_relation, sql) }}
+    {%- endcall %}
+    {% call statement('main') -%}
+        create table {{ intermediate_relation }} as {{ existing_relation }}
+    {%- endcall %}
+    {% call statement('insert_new_data') -%}
+        insert into {{ intermediate_relation }} select * from {{ new_data_relation }}
+    {%- endcall %}
+    {% if execute %}
+      {% set select_changed_partitions %}
+          select distinct {{ partition_by|join(', ') }}
+          from {{ intermediate_relation }}
+      {% endset %}
+      {% set partitions_expressions = run_query(select_changed_partitions).rows %}
+    {% else %}
+      {% set partitions_expressions = [] %}
+    {% endif %}
+    {% for partition_expression in partitions_expressions %}
+      {% call statement('replace_partition_' ~ loop.index) %}
+          alter table {{ existing_relation }}
+          replace partition {{ partition_expression }}
```

Comment:
Also, I have doubts about replacing partitions here in a loop: when multiple partitions are affected, this makes the operation non-atomic: if a subset of partitions is not replaced, the target table might end up being in an inconsistent state. Does dbt have a rollback functionality?

Or does ClickHouse supports manipulating multiple partitions at once? I have not found in documentation and could not compose such a query. Though many Big Data tools (e.g. AWS Athena, Hive) support multiple partitions manipulation when the partitions are listed in a comma-separated fashion.

Reply:
You can actually do it like this
As an example how to do it in jinja you can [look here](https://github.com/ClickHouse/dbt-clickhouse/blob/b79669ae555f5b060ef42dc14a3d6372cd418b33/dbt/include/clickhouse/macros/persist_docs.sql#L13), it does the same with comments array
```sql
ALTER TABLE ... REPLACE PARTITION 1, REPLACE PARTITION 2, REPLACE PARTITION 3
```

But if you have cluster it will not be atomical. One shard can work and other will fail.